_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+      <div className="Drawer-title">
+        {this.props.title}
+      </div>
+
+      <div className="Drawer-content">
+        {this.props.children}
+      </div>
+    </div>
+  );
+}
+```
+
+With a little variable destructing, we can make this even simpler.
+
+``` jsx
+render() {
+  const { title, children } = this.props;
+
+  return (
+    <div className="Drawer">
+      <div className="Drawer-title">
+        {title}
+      </div>
+
+      <div className="Drawer-content">
+        {children}
+      </div>
+    </div>
+  );
+}
+```
+
+Next we need to toggle the above state variable `active` whenever a user clicks the title. We can do this by adding an `onClick` handler to the element.
+
+``` jsx
+// This negates the boolean for active which effectively toggles it between true and false
+handleTitleClick() {
+  this.setState({ active: !this.state.active });
+}
+
+render() {
+  const { title, children } = this.props;
+
+  return (
+    <div className="Drawer">
+      <div className="Drawer-title" onClick={this.handleTitleClick}>
+        {title}
+      </div>
+
+      <div className="Drawer-content">
+        {children}
+      </div>
+    </div>
+  );
+}
+```
+
+Using the active state variable, we can also adjust the CSS classes on the top-level
+HTML element. This will let our stylesheet know the current state of the component.
+
+``` jsx
+render() {
+  const { title, children } = this.props;
+  const drawerStyles = this.state.active ? 'is-expanded' : '';
+
+  return (
+    <div className={`Drawer ${drawerStyles}`}>
+      <div className="Drawer-title" onClick={this.handleTitleClick}>
+        {title}
+      </div>
+
+      <div className="Drawer-content">
+        {children}
+      </div>
+    </div>
+  );
+}
+```
+
+<blockquote class="Info Info-right">
+"setState() does not always immediately update the component. It may batch or defer the update until later. This makes reading this.state right after calling setState() a potential pitfall."
+<cite><a href="https://reactjs.org/docs/react-component.html#setstate">- reactjs.org</a></cite>
+</blockquote>
+
+There's a minor gotcha at this point with the above setting of state. We're using a negated form of the previous state `!this.state.active` to inform the current state what its value is. The problem is that state updates are batched meaning you can't rely on them being immediate. Because of this we need to explictly use the previous state within the `setState` function.
+
+We'll use an arrow function here to pluck the previous state into a variable named `prevState`. This will allow us to be confident that the state change will be accurate.
+
+``` jsx
+handleTitleClick() {
+  this.setState(prevState => ({ active: !prevState.active }));
+};
+```
+
+Now our onClick handler properly deals with existing state.
+
+The onClick handler can only update state by accessing the `this` context. This means we'll also need to bind it to `this` properly within the constructor.
+
```

Comment:
However, the onClick handler can only update state by accessing the `this` context.